File: Derivatives/Option.py
'''@package ForwardFinancialFramework.Derivatives.Option

This package contains the base derivative class, option.
Created on 30 May 2012
'''

class Option:
	"""Base derivative class

	This class represents the base derivative class. In practice, its a simple European future.
	"""
	name = "option"
	##The container for the Underlying upon which this product depends
	underlying = None

	##The defined strike price of produce
	strike_price = 0.0
	##The time period of the product, i.e. the time from the present until expiry
	time_period = 0.0
	##Call or put?
	call = None

	##The next time step required
	delta_time = 0.0
	##The value of the produce
	value = 0.0

	#Model
	#latency_model_coefficients = []
	#accuracy_model_coefficients = []

	def __init__(self,underlying,time_period,call,strike_price):
		"""Constructor

		Parameters
			underlying - (list of FowardFinancialFramework.Underlyings) list of underlyings that this product depends upon
			time_period - (float) time until expiry of product
			call - (bool) call product?
			strike_price - (float) defined expiry price
		"""
		self.underlying = underlying

		self.time_period = time_period
		self.call = call
		self.strike_price = strike_price

		#self.latency_model_coefficients = []
		#self.accuracy_model_coefficients = []

    	"""
    def __repr__(self):
	return_string = ""
	for u in self.underlying:
	    return_string += repr(u)
	return str(self.name + " derivative" + " underlying="+return_string + " time_period=" + str(self.time_period) + " call="+ str(self.call) + " strike_price=" + str(self.strike_price))
    	"""

	def path_init(self):
       		"""Path initialisation method
       		
		Parameters
			None

		Initiate the derivative's path/simulation (resets value back to 0.0)
		"""
		self.value = 0
		self.delta_time = self.time_period
        
	def path(self,price,time):
       		"""Path evolution method
       		
		Parameters
			price - (float) the current price of the underlying product
			time - (time) the current time of the underlying product

		Evolves the derivative's simulation. Is a dummy method for the base class
		"""
   		pass

	def payoff(self,end_price):
		"""Payoff method
		
		Parameters
			end_price - (float) the final price of the underlying

		Finds the value of the product, based upon the end price of the underlying, and if this is a call or not.
		"""
		if(self.call): return end_price - strike_price
		else: return strike_price - end_price
        
"""
    	def latency_model(self,paths):
      value = self.latency_model_coefficients[0]
      for i,coeff in enumerate(self.latency_model_coefficients[1:]): value += coeff*paths**(i+1)
      return value

    def accuracy_model(self,paths):
      value = self.accuracy_model_coefficients[0]
      for i,coeff in enumerate(self.accuracy_model_coefficients[1:]): value += coeff*paths**-(1.0/(i+1))
      return value
"""

File: Platforms/MulticoreCPU/multicore_c_code/option.c
/*
 * option.c
 *
 *  Created on: 10 November 2012
 *      Author: gordon
 */
#include "option.h"

void option_derivative_init(FP_t t,FP_t c,FP_t k,option_attributes* o_a){
	o_a->strike_price = k;
	o_a->call=c;
	o_a->time_period=t;
	//o_a->points = p;
}

void option_derivative_path_init(option_variables* o_v,option_attributes* o_a){
	o_v->value=0;
	o_v->delta_time=o_a->time_period;
	//o_v->delta_time=o_a->time_period/o_a->points;
}

void option_derivative_path(FP_t price,FP_t time,option_variables* o_v,option_attributes* o_a){
}

void option_derivative_payoff(FP_t end_price,option_variables* o_v,option_attributes* o_a){
	if((o_a->call)){
		o_v->value = end_price - o_a->strike_price;
	}
	else{
		o_v->value = o_a->strike_price - end_price;
	}
}
